src/ssl.c

```diff
@@ -258,6 +260,533 @@ const WOLF_EC_NIST_NAME kNistCurves[] = {
+
+    decodedConfigs[decodedLen - 1] = 0;
+
+    /* decode the echConfigs */
+    ret = Base64_Decode((byte*)echConfigs64, echConfigs64Len,
+      decodedConfigs, &decodedLen);
+
+    if (ret != 0) {
+        XFREE(decodedConfigs, ssl->heap, DYNAMIC_TYPE_TMP_BUFFER);
+        return ret;
+    }
+
+    ret = wolfSSL_SetEchConfigs(ssl, decodedConfigs, decodedLen);
+
+    XFREE(decodedConfigs, ssl->heap, DYNAMIC_TYPE_TMP_BUFFER);
+
+    return ret;
+}
+
+int wolfSSL_SetEchConfigs(WOLFSSL* ssl, const byte* echConfigs,
+  word32 echConfigsLen)
+{
+    int i;
+    int j;
+    word16 version;
+    word16 length;
+    word16 hpkePubkeyLen;
+    word16 cipherSuitesLen;
+    word16 publicNameLen;
+    WOLFSSL_EchConfig* configList = NULL;
+    WOLFSSL_EchConfig* workingConfig = NULL;
+    WOLFSSL_EchConfig* lastConfig = NULL;
+    byte* echConfig = NULL;
+
+    if (ssl == NULL || echConfigs == NULL || echConfigsLen == 0)
+        return BAD_FUNC_ARG;
+
+    /* already have ech configs */
+    if (ssl->options.useEch == 1) {
+        return WOLFSSL_FATAL_ERROR;
+    }
+
+    /* check that the total length is well formed */
+    if (XNTOHS(((word16*)echConfigs)[0]) == echConfigsLen - 2)
+    {
+        /* skip the total length uint16_t */
+        i = 2;
+
+        do {
+            echConfig = (byte*)echConfigs + i;
+            ato16(echConfig, &version);
+            ato16(echConfig + 2, &length);
+
+            /* if the version does not match */
+            if (version != TLSX_ECH) {
+                /* we hit the end of the configs */
+                if ( (word32)i + 2 >= echConfigsLen ) {
+                    break;
+                }
+
+                /* skip this config, +4 for version and length */
+                i += length + 4;
+                continue;
+            }
+
+            if (workingConfig == NULL) {
+                workingConfig =
+                    (WOLFSSL_EchConfig*)XMALLOC(sizeof(WOLFSSL_EchConfig),
+                    ssl->heap, DYNAMIC_TYPE_TMP_BUFFER);
+                configList = workingConfig;
+                workingConfig->next = NULL;
+            }
+            else {
+                lastConfig = workingConfig;
+                workingConfig->next =
+                    (WOLFSSL_EchConfig*)XMALLOC(sizeof(WOLFSSL_EchConfig),
+                    ssl->heap, DYNAMIC_TYPE_TMP_BUFFER);
+                workingConfig = workingConfig->next;
+            }
+
+            ForceZero(workingConfig, sizeof(WOLFSSL_EchConfig));
+
+            /* rawLen */
+            workingConfig->rawLen = length + 4;
+
+            /* raw body */
+            workingConfig->raw = (byte*)XMALLOC(workingConfig->rawLen,
+                ssl->heap, DYNAMIC_TYPE_TMP_BUFFER);
+            XMEMCPY(workingConfig->raw, echConfig, workingConfig->rawLen);
+
+            /* skip over version and length */
+            echConfig += 4;
+
+            /* configId, 1 byte */
+            workingConfig->configId = *(echConfig);
```

Comment:
Extracting data from a variable length message and I can't see checks for running over message length.